adapters/play/src/test/scala/caliban/AdapterSpec.scala

```diff
@@ -0,0 +1,221 @@
+    ZIO.collectAllPar(for {
+      file <- files
+    } yield {
+      for {
+        bytes <- file.allBytes
+        meta  <- file.meta
+      } yield TestAPI.File(
+        Service.hex(Service.sha256(bytes.toArray)),
+        meta.map(_.path.toAbsolutePath.toString).getOrElse(""),
+        meta.map(_.fileName).getOrElse(""),
+        meta.flatMap(_.contentType).getOrElse("")
+      )
+    })
+
+  def sha256(b: Array[Byte]) =
+    MessageDigest.getInstance("SHA-256").digest(b)
+
+  def hex(b: Array[Byte]): String =
+    String.format("%032x", new BigInteger(1, b))
+}
+
+case class UploadFileArgs(file: Upload)
+case class UploadFilesArgs(files: List[Upload])
+
+object TestAPI extends GenericSchema[Blocking with Uploads with Console with Clock] {
+  val api: GraphQL[Blocking with Uploads with Console with Clock] =
+    graphQL(
+      RootResolver(
+        Queries(args => UIO("stub")),
+        Mutations(args => Service.uploadFile(args.file), args => Service.uploadFiles(args.files))
+      )
+    )
+
+  implicit val uploadFileArgsSchema = gen[UploadFileArgs]
+  implicit val mutationsSchema      = gen[Mutations]
+  implicit val queriesSchema        = gen[Queries]
+
+  case class File(hash: String, path: String, filename: String, mimetype: String)
+
+  case class Queries(stub: Unit => UIO[String])
+
+  case class Mutations(
+    uploadFile: UploadFileArgs => ZIO[Blocking with Uploads, Throwable, File],
+    uploadFiles: UploadFilesArgs => ZIO[Blocking with Uploads, Throwable, List[File]]
+  )
+}
+
+object AdapterSpec extends DefaultRunnableSpec {
+  val httpCliLayer: ZLayer[Any, Throwable, HttpClient] =
+    HttpClient.makeHttpClient.toLayer >>> HttpClient.make("127.0.0.1", 8088)
+  val interpLayer: ZLayer[Any, Any, Console with Clock] = Console.live ++ Clock.live
+  val runtime: Runtime[Console with Clock with Blocking with Random with Uploads] =
+    Runtime.unsafeFromLayer(
+      Console.live ++ Clock.live ++ Blocking.live ++ Random.live ++ Uploads.Service.empty,
+      Platform.default
+    )
+
+  // THIS BLOCKS INDEFINTELY DUE TO httpCliLayer
+  //  val runtime: Runtime[HService with HttpClient with Console with Clock] =
+  //    Runtime.unsafeFromLayer(
+  //      (Service.make ++ httpCliLayer ++ Console.live ++ Clock.live),
+  //      Platform.default
+  //    )
+
+  val interpreter: GraphQLInterpreter[Blocking with Uploads with Console with Clock, CalibanError] =
+    runtime.unsafeRun(TestAPI.api.interpreter)
+  val server = AkkaHttpServer.fromRouterWithComponents(
+    ServerConfig(
+      mode = Mode.Dev,
+      port = Some(8088),
+      address = "127.0.0.1"
+    )
+  ) { components =>
+    PlayRouter(
+      interpreter,
+      DefaultControllerComponents(
+        components.defaultActionBuilder,
+        components.playBodyParsers,
+        components.messagesApi,
+        components.langs,
+        components.fileMimeTypes,
+        components.executionContext
+      )
+    )(runtime, components.materializer).routes
+  }
```

Comment:
You can package this up into a `Layer` then provide it to the `Spec` via a `provideSomeLayerShared`

Reply:
Expanding on this you can do something like this instead:

```scala
val apiLayer = ZLayer.fromEffect(for {
  interpreter <- TestAPI.api.interpreter
} yield AkkaHttpServer.fromRouterWithComponents(
    ServerConfig(
      mode = Mode.Dev,
      port = Some(8088),
      address = "127.0.0.1"
    )
  ) { components =>
    PlayRouter(
      interpreter,
      DefaultControllerComponents(
        components.defaultActionBuilder,
        components.playBodyParsers,
        components.messagesApi,
        components.langs,
        components.fileMimeTypes,
        components.executionContext
      )
    )(runtime, components.materializer).routes
  }


val specLayer = AsyncHttpZioBackend.layer() ++ ((Uploads.Service.empty ++  ZEnv.any) >>> apiLayer)


suite("Requests")(...).provideCustomLayerShared[ZEnv](specLayer)
```

can use `fromManaged` as well, if you want to provide a shutdown hook to terminate the `AkkaServer`